Firestore/core/src/firebase/firestore/local/leveldb_persistence.mm

```diff
@@ -0,0 +1,284 @@
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/leveldb_persistence.h"
+
+#include <limits>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_lru_reference_delegate.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_migrations.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_util.h"
+#include "Firestore/core/src/firebase/firestore/local/listen_sequence.h"
+#include "Firestore/core/src/firebase/firestore/local/lru_garbage_collector.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_delegate.h"
+#include "Firestore/core/src/firebase/firestore/local/sizer.h"
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/string_util.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/match.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+namespace {
+
+using auth::User;
+using leveldb::DB;
+using model::ListenSequenceNumber;
+using util::Path;
+using util::Status;
+using util::StatusOr;
+using util::StringFormat;
+
+/**
+ * Finds all user ids in the database based on the existence of a mutation
+ * queue.
+ */
+std::set<std::string> CollectUserSet(LevelDbTransaction* transaction) {
+  std::set<std::string> result;
+
+  std::string tablePrefix = LevelDbMutationKey::KeyPrefix();
+  auto it = transaction->NewIterator();
+  it->Seek(tablePrefix);
+
+  LevelDbMutationKey rowKey;
+  while (it->Valid() && absl::StartsWith(it->key(), tablePrefix) &&
+         rowKey.Decode(it->key())) {
+    result.insert(rowKey.user_id());
+
+    auto userEnd = LevelDbMutationKey::KeyPrefix(rowKey.user_id());
+    userEnd = util::PrefixSuccessor(userEnd);
+    it->Seek(userEnd);
+  }
+  return result;
+}
+
+}  // namespace
+
+util::StatusOr<std::unique_ptr<LevelDbPersistence>> LevelDbPersistence::Create(
+    util::Path dir,
+    FSTLocalSerializer* serializer,
+    const LruParams& lru_params) {
+  Status status = EnsureDirectory(dir);
+  if (!status.ok()) return status;
+
+  status = ExcludeFromBackups(dir);
+  if (!status.ok()) return status;
+
+  StatusOr<std::unique_ptr<DB>> created = OpenDb(dir);
+  if (!created.ok()) return created.status();
+
+  std::unique_ptr<DB> db = std::move(created).ValueOrDie();
+  LevelDbMigrations::RunMigrations(db.get());
+
+  LevelDbTransaction transaction(db.get(), "Start LevelDB");
+  std::set<std::string> users = CollectUserSet(&transaction);
+  transaction.Commit();
+
+  // Explicitly conversion is required to allow the StatusOr to be created.
+  std::unique_ptr<LevelDbPersistence> result(new LevelDbPersistence(
```

Comment:
That doesn't work here because the constructor is private.

I've tried other fixes for this, like creating a private subclass of this that forces the constructor to be public and instantiating that here but those cures are worse than the disease.